File: src/formats/moment.js
import moment from 'moment';
import { nulFormat } from '../globals';

const DateFormatterDefault = 'YYYY-MM-DD';
const DateTimeFormatterDefault = 'YYYY-MM-DD h:mm a';
const MonthAndDayFormatterDefault = 'MM/DD';
const TimeFormatterDefault = 'h:mm a';

moment.suppressDeprecationWarnings = true;
const _minDateValue = moment([1753, 1, 1]);
const _maxDateValue = moment([3000, 1, 1]);

// date
export const dateFormatter = (value, param) => {
  if (!value) return nulFormat;
  value = moment(value);
  if (param && param.format) {
    switch (param.format) {
      case 'date': return value.format('DD MMMM YYYY');
      case 'longDate': return value.format('dddd, MMMM D, YYYY');
      case 'longDate2': return value.format(value.year() === moment().year() ? 'dddd, MMMM D' : 'dddd, MMMM D, YYYY');
      case 'shortDate': return value.format('D-MMM-YYYY');
      case 'shorterDate': return value.format('MMM D YYYY');
      case 'monthDay': return value.format('MMMM D');
      case 'monthYear': return value.format('MMMM YYYY');
      case 'pattern': return value.format(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.format(DateFormatterDefault);
};
export const dateParser = (text, param, error) => {
  if (!text) return [text, true, error];
  let value = getDate(text); if (!value.isValid()) return [text, false, error];
  if (value.isBefore(_minDateValue) || value.isAfter(_maxDateValue)) return [text, false, error];
  value = moment([value.year(), value.month(), value.date()]);
  if (param) { // check param
    const minValue = param.minValue; if (minValue && moment(minValue).isAfter(value)) return [text, false, error];
    const maxValue = param.maxValue; if (maxValue && value.isAfter(moment(maxValue))) return [text, false, error];
  }
  return [value, true, error];
};

// dateTime
export const dateTimeFormatter = (value, param) => {
  if (!value) return nulFormat;
  value = moment(value);
  if (param && param.format) {
    switch (param.format) {
      case 'dateTime': return value.format('DD MMMM YYYY h:mm a');
      case 'longDateTime': return value.format('dddd, MMMM D, YYYY h:mm a');
      case 'longDate': return value.format('dddd, MMMM D, YYYY');
      case 'longTime': return value.format('hh:mm:ss a');
      case 'shortDate': return value.format('D-MMM-YYYY');
      case 'shorterDate': return value.format('MMM D YYYY');
      case 'shortTime': return value.format('h:mm a');
      case 'tinyDate': return value.format('M/D/YY');
      case 'tinyDateTime': return value.format('M/D/YY h:mm a');
      case 'pattern': return value.format(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.format(DateTimeFormatterDefault);
};
export const dateTimeParser = (text, param, error) => {
  if (!text) return [text, true, error];
  let value = getDateAndTime(text); if (!value.isValid()) return [text, false, error];
  if (value.isBefore(_minDateValue) || value.isAfter(_maxDateValue)) return [text, false, error];
  if (param) { // check param
    const minValue = param.minValue; if (minValue && moment(minValue).isAfter(value)) return [text, false, error];
    const maxValue = param.maxValue; if (maxValue && value.isAfter(moment(maxValue))) return [text, false, error];
  }
  return [value, true, error];
};

// monthAndDay
export const monthAndDayFormatter = (value, param) => {
  if (!value) return nulFormat;
  value = moment(value);
  if (param) {
    switch (param.format) {
      case 'pattern': return value.format(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.format(MonthAndDayFormatterDefault);
};
export const monthAndDayParser = (text, param, error) => {
  if (!text) return [text, true, error];
  const match = /^((0[1-9])|(1[0-2]))[\/-](([0-2][0-9])|([3][0-1]))$/.exec(text); if (!match) return [text, false, error];
  let value = moment().set({ year: 2000, month: match[1] - 1, date: match[4], h: 0, m: 0, s: 0 }); //if (!value.isValid()) return [text, false, error];
  return [value, true, error];
};

// time
export const timeFormatter = (value, param) => {
  if (!value) return nulFormat;
  value = moment(value);
  if (param && param.format) {
    switch (param.format) {
      case 'longTime': return value.format('h:mm:ss a');
      case 'shortTime': return value.format('h:mm a');
      case 'pattern': return value.format(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.format(TimeFormatterDefault);
};
export const timeParser = (text, param, error) => {
  if (!text) return [text, true, error];
  let value = getTime(text); if (!value.isValid()) return [text, false, error];
  value = moment({ year: 2000, month: 0, date: 1, h: value.hour(), m: value.minute(), s: value.second() });
  if (param) { // check param
    const minValue = param.minValue; if (minValue) {
      let theValue = moment(minValue, 'hh:mm:ss a');
      theValue = moment({ year: 2000, month: 0, date: 1, h: theValue.hour(), m: theValue.minute(), s: theValue.second() });
      if (theValue.isAfter(value)) return [text, false, error];
    }
    const maxValue = param.maxValue; if (maxValue) {
      let theValue = moment(maxValue, 'hh:mm:ss a');
      theValue = moment({ year: 2000, month: 0, date: 1, h: theValue.hour(), m: theValue.minute(), s: theValue.second() });
      if (value.isAfter(theValue)) return [text, false, error];
    }
  }
  return [value, true, error];
};


const getDate = (text) => moment(text);

const getDateAndTime = (text) => {
  let value;
  if ((value = moment(text)).isValid()) return value;
  if ((value = moment(text, 'YYYY-MM-DDTHH:mm')).isValid()) return value;
  return value;
}

const getTime = (text) => {
  let value;
  if (!text.includes(':')) return moment('bad');
  if ((value = moment(text)).isValid()) return value;
  /* istanbul ignore next */
  if ((value = moment(text, 'YYYY-MM-DDTHH:mm')).isValid()) return value;
  /* istanbul ignore next */
  if ((value = moment(text, 'hh:mm:ss a')).isValid()) return value;
  /* istanbul ignore next */
  return value;
}

File: src.cs/CoreValidation/Formats/Moment.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text.RegularExpressions;
using static CoreValidation.Globals;

namespace CoreValidation.Formats
{
  public static class Moment
  {
    const string DateFormatterDefault = "yyyy-MM-dd";
    const string DateTimeFormatterDefault = "yyyy-MM-dd h:mm tt";
    const string MonthAndDayFormatterDefault = "MM/dd";
    const string TimeFormatterDefault = "h:mm tt";

    static readonly DateTime _minDateValue = new DateTime(1753, 1, 1);
    static readonly DateTime _maxDateValue = new DateTime(3000, 1, 1);
    static readonly Regex _monthAndDayPattern = new Regex($"^((0[1-9])|(1[0-2]))[\\/-](([0-2][0-9])|([3][0-1]))$", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline);

    // date
    public static string DateFormatter(object value, IDictionary<string, object> param = null)
    {
      if (value == null || (value is string h && h.Length == 0)) return NulFormat;
      var value_ = Convert.ToDateTime(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "date": return value_.ToString("dd MMMM yyyy", CultureInfo.InvariantCulture);
          case "longDate": return value_.ToString("dddd, MMMM d, yyyy", CultureInfo.InvariantCulture);
          case "longDate2": return value_.ToString(value_.Year == DateTime.Today.Year ? "dddd, MMMM d" : "dddd, MMMM d, yyyy", CultureInfo.InvariantCulture);
          case "shortDate": return value_.ToString("d-MMM-yyyy", CultureInfo.InvariantCulture);
          case "shorterDate": return value_.ToString("MMM d yyyy", CultureInfo.InvariantCulture);
          case "monthDay": return value_.ToString("MMMM d", CultureInfo.InvariantCulture);
          case "monthYear": return value_.ToString("MMMM yyyy", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString(DateFormatterDefault, CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) DateParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!TryDate(text, out var value)) return (text, false, error);
      if (value < _minDateValue || value > _maxDateValue) return (text, false, error);
      value = new DateTime(value.Year, value.Month, value.Day);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is DateTime minValue) { if (value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is DateTime maxValue) { if (value > maxValue) return (text, false, error); }
      }
      return (value, true, error);
    }

    // dateTime
    public static string DateTimeFormatter(object value, IDictionary<string, object> param = null)
    {
      if (value == null || (value is string h && h.Length == 0)) return NulFormat;
      var value_ = Convert.ToDateTime(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "dateTime": return value_.ToString("dd MMMM yyyy h:mm tt", CultureInfo.InvariantCulture);
          case "longDateTime": return value_.ToString("dddd, MMMM d, yyyy h:mm tt", CultureInfo.InvariantCulture);
          case "longDate": return value_.ToString("dddd, MMMM d, yyyy", CultureInfo.InvariantCulture);
          case "longTime": return value_.ToString("hh:mm:ss tt", CultureInfo.InvariantCulture);
          case "shortDate": return value_.ToString("d-MMM-yyyy", CultureInfo.InvariantCulture);
          case "shorterDate": return value_.ToString("MMM d yyyy", CultureInfo.InvariantCulture);
          case "shortTime": return value_.ToString("h:mm tt", CultureInfo.InvariantCulture);
          case "tinyDate": return value_.ToString("M/d/yy", CultureInfo.InvariantCulture);
          case "tinyDateTime": return value_.ToString("M/d/yy h:mm tt", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString(DateTimeFormatterDefault, CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) DateTimeParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!TryDateAndTime(text, out var value)) return (text, false, error);
      if (value < _minDateValue || value > _maxDateValue) return (text, false, error);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is DateTime minValue) { if (value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is DateTime maxValue) { if (value > maxValue) return (text, false, error); }
      }
      return (value, true, error);
    }

    // monthAndDay
    public static string MonthAndDayFormatter(object value, IDictionary<string, object> param = null)
    {
      if (value == null || (value is string h && h.Length == 0)) return NulFormat;
      var value_ = Convert.ToDateTime(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString(MonthAndDayFormatterDefault, CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) MonthAndDayParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      var match = _monthAndDayPattern.Match(text).Groups; if (match.Count != 7) return (text, false, error);
      var value = new DateTime(2000, int.Parse(match[1].Value), int.Parse(match[4].Value));
      return (value, true, error);
    }

    // time
    public static string TimeFormatter(object value, IDictionary<string, object> param = null)
    {
      if (value == null || (value is string h && h.Length == 0)) return NulFormat;
      var value_ = Convert.ToDateTime(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "longTime": return value_.ToString("h:mm:ss tt", CultureInfo.InvariantCulture);
          case "shortTime": return value_.ToString("h:mm tt", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString(TimeFormatterDefault, CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) TimeParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!TryTime(text, out var value)) return (text, false, error);
      value = new DateTime(2000, 1, 1, value.Hour, value.Minute, value.Second);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is DateTime minValue) { if (value < new DateTime(2000, 1, 1, minValue.Hour, minValue.Minute, minValue.Second)) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is DateTime maxValue) { if (value > new DateTime(2000, 1, 1, maxValue.Hour, maxValue.Minute, maxValue.Second)) return (text, false, error); }
      }
      return (value, true, error);
    }

    static bool TryDate(string text, out DateTime value)
    {
      if (
        //text.IndexOf(":") != -1 ||
        //text.IndexOf("am", StringComparison.OrdinalIgnoreCase) != -1 ||
        //text.IndexOf("pm", StringComparison.OrdinalIgnoreCase) != -1 ||
        !DateTime.TryParse(text, out var z))
      {
        value = default; return false;
      }
      value = z.Date; return true;
    }

    static bool TryDateAndTime(string text, out DateTime value) => DateTime.TryParse(text, out value);

    static bool TryTime(string text, out DateTime value)
    {
      if (text.IndexOf(":") == -1 ||
        !DateTime.TryParse(text, out var z))
      {
        value = default; return false;
      }
      value = z; return true;
    }
  }
}
